### PR TITLE
Feat: Support SSH for retrieving terraform modules in private git repo

### PR DESCRIPTION
**Pull request by motilayo**
### Support SSH for retrieving terraform modules in private git repo

Enable support for terraform modules in private git repo through SSH. A kubernetes secret which holds the ssh private-key and known_hosts should be created and referenced in the component definition.

Terraform controller PR: https://github.com/kubevela/terraform-controller/pull/349

Feature for https://github.com/kubevela/terraform-controller/issues/292

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [x] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Tested in a local cluster. 
1. Create a kubernetes secret of type `kubernetes.io/ssh-auth` for the git ssh credentials & known-hosts
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: git-ssh
  namespace: vela-system
type: kubernetes.io/ssh-auth
stringData:
  ssh-privatekey: |
   <SSH Private Key> # the ssh private key used for authenticating git
  known_hosts: |
   <SSH known_hosts>  # use `ssh-keyscan github.com`  to generate known_hosts
```

2. Create a component definition
``` yaml
apiVersion: core.oam.dev/v1beta1
kind: ComponentDefinition
metadata:
  annotations:
    definition.oam.dev/description: Terraform module to provision a KMS key with alias
  labels:
    type: terraform
  name: aws-kms-key
  namespace: vela-system
spec:
  schematic:
    terraform:
      configuration: git@github.com:motilayo/terraform-aws-kms-key.git # use terraform configuration in private git repo to test
      gitCredentialsSecretReference: # reference to the secret created earlier
        name: git-ssh # secret name
        namespace: vela-system  # secret namespace
      type: remote
```

3. Create an application and verify

### CODE REVIEW

**Comment by motilayo** [comment]
@wonderflow Please review when you get a chance. I've added some tests and rebased my branch with the master branch.

File: go.mod [hunk 1]
```diff
@@ -116,6 +117,8 @@ require (
 	sigs.k8s.io/yaml v1.3.0
 )
 
+require github.com/go-git/go-git/v5 v5.5.1
```
Comment (Somefive) [inline go.mod:120]: move it into L118

File: pkg/controller/utils/capability.go [hunk 2]
```diff
@@ -25,15 +25,20 @@ import (
 	"strings"
 
 	"github.com/getkin/kin-openapi/openapi3"
+
+	"github.com/go-git/go-git/v5"
+	gitssh "github.com/go-git/go-git/v5/plumbing/transport/ssh"
+
```
Comment (Somefive) [inline pkg/controller/utils/capability.go:31]: remove L28 & L31

File: pkg/controller/utils/capability.go [hunk 3]
```diff
@@ -25,15 +25,20 @@ import (
...
 	"github.com/pkg/errors"
-	"gopkg.in/src-d/go-git.v4"
 	v1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/klog/v2"
 	"k8s.io/utils/pointer"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
+	k8stypes "k8s.io/apimachinery/pkg/types"
```
Comment (Somefive) [inline pkg/controller/utils/capability.go:40]: remove L39

**Review by Somefive — COMMENTED** [review]
LGTM after nit.

File: pkg/appfile/appfile.go [hunk 4]
```diff
@@ -724,6 +724,10 @@ func generateTerraformConfigurationWorkload(wl *Workload, ns string) (*unstructu
 		configuration.Spec.ProviderReference = wl.FullTemplate.ComponentDefinition.Spec.Schematic.Terraform.ProviderReference
 	}
 
+	if configuration.Spec.GitCredentialsSecretReference == nil {
+		configuration.Spec.GitCredentialsSecretReference = wl.FullTemplate.ComponentDefinition.Spec.Schematic.Terraform.GitCredentialsSecretReference
```
Comment (wonderflow) [inline pkg/appfile/appfile.go:728]: what if the value is still nil?

**Review by wonderflow — APPROVED** [review]
great job, generally LGTM!

File: go.mod [hunk 5]
```diff
@@ -64,7 +64,8 @@ require (
 	github.com/oam-dev/cluster-gateway v1.4.0
 	github.com/oam-dev/cluster-register v1.0.4-0.20220928064144-5f76a9d7ca8c
 	github.com/oam-dev/terraform-config-inspect v0.0.0-20210418082552-fc72d929aa28
-	github.com/oam-dev/terraform-controller v0.7.0
+	// github.com/oam-dev/terraform-controller v0.7.0
+	github.com/oam-dev/terraform-controller v0.7.8-0.20221209023130-841b0d17d9fa
```
[the remark this answers is not in the document]
Comment (chivalryq) [reply go.mod:68]: Yes, new version released, please use v0.7.8 now.

File: pkg/appfile/appfile.go (hunk 4 above)
Comment (chivalryq) [reply pkg/appfile/appfile.go:728]: Then it's ok to have no credentials. Like when it is a public repo.

File: pkg/appfile/appfile.go [hunk 6]
```diff
@@ -724,6 +724,10 @@ func generateTerraformConfigurationWorkload(wl *Workload, ns string) (*unstructu
...
+	}
+
 	switch wl.FullTemplate.Terraform.Type {
 	case "hcl":
 		configuration.Spec.HCL = wl.FullTemplate.Terraform.Configuration
```
Comment (chivalryq) [inline pkg/appfile/appfile.go:733]: See L745-L748
```
	delete(variableMap, WriteConnectionSecretToRefKey)
	delete(variableMap, RegionKey)
	delete(variableMap, ProviderRefKey)
	delete(variableMap, ForceDeleteKey)
```
We should also delete GitCredentialsSecretReference Key here.

File: pkg/controller/utils/capability.go [hunk 7]
```diff
@@ -347,6 +361,45 @@ func GetKubeSchematicOpenAPISchema(params []commontypes.KubeParameter) ([]byte,
 	return generateJSONSchemaWithRequiredProperty(properties, required)
 }
 
+// GetGitSSHPublicKey gets a kubernetes secret containing the SSH private key based on  based on gitCredentialsSecretReference parameters for component and trait definition
```
Comment (chivalryq) [inline pkg/controller/utils/capability.go:364]: "based on  based on" typo

File: pkg/controller/utils/capability.go [hunk 8]
```diff
@@ -347,6 +361,45 @@ func GetKubeSchematicOpenAPISchema(params []commontypes.KubeParameter) ([]byte,
...
+	sshPrivateKey := secret.Data[v1.SSHAuthPrivateKey]
+	publicKey, err := gitssh.NewPublicKeys("git", sshPrivateKey, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate public key from private key: %w", err)
+	}
+	sshKnownHosts := secret.Data[GitCredsKnownHosts]
+	sshDir := filepath.Join(os.TempDir(), ".ssh")
+	sshKnownHostsPath := filepath.Join(sshDir, GitCredsKnownHosts)
+	_ = os.Mkdir(sshDir, 0700)
+	err = os.WriteFile(sshKnownHostsPath, sshKnownHosts, 0600)
```
Comment (chivalryq) [inline pkg/controller/utils/capability.go:394]: We should clean this file after using it.

File: pkg/controller/utils/capability_test.go [hunk 9]
```diff
@@ -386,3 +395,143 @@ variable "bbb" {
...
+		name string
+		args args
+		want want
+	}{
+		{
+			name: "git credentials secret not found",
+			args: args{
+				k8sClient: k8sClient,
+				GitCredentialsSecretReference: &corev1.SecretReference{
+					Name:      "git-shh-auth",
```
Comment (chivalryq) [inline pkg/controller/utils/capability_test.go:467]: This case can be a little illegible, let's make the name more explicit, like `auth-secret-not-exist`

File: pkg/controller/utils/capability_test.go [hunk 10]
```diff
@@ -460,7 +460,7 @@ func TestGetGitSSHPublicKey(t *testing.T) {
 		want want
 	}{
 		{
-			name: "git credentials secret not found",
+			name: "git credentials secret does not exist",
 			args: args{
 				k8sClient: k8sClient,
 				GitCredentialsSecretReference: &corev1.SecretReference{
```
Comment (chivalryq) [inline pkg/controller/utils/capability_test.go:466]: Sorry for misleading, I mean the secret's Name. `shh` is kind of illegible :D

**Review by chivalryq — COMMENTED** [review]
LGTM! Please fix tests.

**Pull request merged**